File: android/src/test/kotlin/com/vaam/flutter_sign_keypair/WireContractTest.kt
package com.vaam.flutter_sign_keypair

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * The wire contract, pinned on the Kotlin side.
 *
 * The same vectors are asserted in `test/wire_contract_test.dart` and
 * `darwin_tests/.../WireContractTests.swift`. The mapping is necessarily
 * duplicated once per language, and duplication is what drifts — these tests
 * are what turn a drift into a red build on the side that moved, instead of a
 * runtime failure on one platform only.
 */
class WireContractTest {

    // --- SignerMethod -------------------------------------------------------------

    @Test
    fun `method wire names match the Dart side verbatim`() {
        val expected = mapOf(
            "capabilities" to SignerMethod.CAPABILITIES,
            "generateKey" to SignerMethod.GENERATE_KEY,
            "getKey" to SignerMethod.GET_KEY,
            "sign" to SignerMethod.SIGN,
            "deleteKey" to SignerMethod.DELETE_KEY,
        )
        for ((wire, method) in expected) {
            assertEquals(wire, method.wire, "wire name for $method")
            assertEquals(method, SignerMethod.from(wire), "parse of \"$wire\"")
        }
        assertEquals(
            expected.values.toSet(),
            SignerMethod.entries.toSet(),
            "a SignerMethod was added without a wire vector — update Wire.kt, " +
                "models.dart, Wire.swift and all three test suites",
        )
    }

    @Test
    fun `every method round-trips`() {
        for (method in SignerMethod.entries) {
            assertEquals(method, SignerMethod.from(method.wire))
        }
    }

    @Test
    fun `method wire names are unique`() {
        assertEquals(SignerMethod.entries.size, SignerMethod.entries.map { it.wire }.toSet().size)
    }

    /**
     * An unknown method must parse to null so the plugin can answer
     * `notImplemented()` — the response Flutter expects, and the one that lets a
     * newer Dart side probe for a capability this build does not have.
     */
    @Test
    fun `unknown method names parse to null`() {
        for (wire in listOf("", " ", "nope", "CAPABILITIES", "generate_key", "sign ")) {
            assertNull(SignerMethod.from(wire), "\"$wire\" should not parse")
        }
    }

    // --- KeyBacking ------------------------------------------------------------

    @Test
    fun `backing wire names match the Dart side verbatim`() {
        val expected = mapOf(
            "strongbox" to KeyBacking.STRONGBOX,
            "tee" to KeyBacking.TEE,
            "software" to KeyBacking.SOFTWARE,
        )
        for ((wire, backing) in expected) {
            assertEquals(wire, backing.wire)
            assertEquals(backing, KeyBacking.from(wire))
        }
        assertEquals(expected.values.toSet(), KeyBacking.entries.toSet())
    }

    /**
     * Android never emits these two, but Dart understands them because iOS does.
     * Pinned here so nobody "tidies up" the Dart enum by deleting them.
     */
    @Test
    fun `android does not emit the apple backing tags`() {
        assertNull(KeyBacking.from("secure_enclave"))
        assertNull(KeyBacking.from("keychain"))
    }

    @Test
    fun `every backing round-trips`() {
        for (backing in KeyBacking.entries) {
            assertEquals(backing, KeyBacking.from(backing.wire))
        }
    }

    @Test
    fun `backing wire names are unique`() {
        assertEquals(KeyBacking.entries.size, KeyBacking.entries.map { it.wire }.toSet().size)
    }

    @Test
    fun `unknown backing tags parse to null and are never treated as hardware`() {
        for (wire in listOf("", "quantum", "STRONGBOX", "strong_box", "tee ")) {
            val parsed = KeyBacking.from(wire)
            assertNull(parsed, "\"$wire\" should not parse")
            // The caller's fallback for a null parse is SOFTWARE — fail-safe.
            assertFalse((parsed ?: KeyBacking.SOFTWARE).isHardwareBacked)
        }
    }

    @Test
    fun `only secure-element backings are hardware-backed`() {
        assertTrue(KeyBacking.STRONGBOX.isHardwareBacked)
        assertTrue(KeyBacking.TEE.isHardwareBacked)
        assertFalse(KeyBacking.SOFTWARE.isHardwareBacked)
    }

    // --- KeyProtection ------------------------------------------------------------

    @Test
    fun `protection wire names match the Dart side verbatim`() {
        val expected = mapOf(
            "ambient" to KeyProtection.AMBIENT,
            "user_present" to KeyProtection.USER_PRESENT,
        )
        for ((wire, protection) in expected) {
            assertEquals(wire, protection.wire)
            assertEquals(protection, KeyProtection.from(wire))
        }
        assertEquals(
            expected.values.toSet(),
            KeyProtection.entries.toSet(),
            "a KeyProtection was added without a wire vector — update Wire.kt, " +
                "models.dart, Wire.swift and all three test suites",
        )
    }

    @Test
    fun `every protection round-trips`() {
        for (protection in KeyProtection.entries) {
            assertEquals(protection, KeyProtection.from(protection.wire))
        }
    }

    @Test
    fun `protection wire names are unique`() {
        assertEquals(
            KeyProtection.entries.size,
            KeyProtection.entries.map { it.wire }.toSet().size,
        )
    }

    /**
     * Unlike [KeyBacking], an unparseable protection tag has **no safe
     * fallback** — the caller must fail. Defaulting to AMBIENT would hand a
     * tier-3 caller a key that signs without a human; defaulting to USER_PRESENT
     * would make the request interceptor prompt on every background poll. The
     * policy is baked into the key permanently at creation, so getting it wrong
     * is not recoverable without destroying the key.
     */
    @Test
    fun `unknown protection tags parse to null`() {
        for (wire in listOf("", " ", "AMBIENT", "userPresent", "user-present", "user_present ")) {
            assertNull(KeyProtection.from(wire), "\"$wire\" should not parse")
        }
    }

    @Test
    fun `only the user-present key demands a human`() {
        assertFalse(KeyProtection.AMBIENT.requiresUserPresence)
        assertTrue(KeyProtection.USER_PRESENT.requiresUserPresence)
    }

    // --- SignerErrorCode ----------------------------------------------------------

    @Test
    fun `error code wire names match the Dart side verbatim`() {
        val expected = mapOf(
            "key_not_found" to SignerErrorCode.KEY_NOT_FOUND,
            "key_already_exists" to SignerErrorCode.KEY_ALREADY_EXISTS,
            "hardware_unavailable" to SignerErrorCode.HARDWARE_UNAVAILABLE,
            "keystore_failure" to SignerErrorCode.KEYSTORE_FAILURE,
            "user_authentication_required" to SignerErrorCode.USER_AUTHENTICATION_REQUIRED,
            "user_authentication_cancelled" to SignerErrorCode.USER_AUTHENTICATION_CANCELLED,
            "key_invalidated" to SignerErrorCode.KEY_INVALIDATED,
        )
        for ((wire, code) in expected) {
            assertEquals(wire, code.wire)
            assertEquals(code, SignerErrorCode.from(wire))
        }
        assertEquals(expected.values.toSet(), SignerErrorCode.entries.toSet())
    }

    /**
     * Dart has two extra members. `unsupported_platform` comes from
     * MissingPluginException and `unknown` is the catch-all, so neither is ever
     * emitted from here. Asserted so a future Kotlin addition of either is
     * caught as the contract change it would be.
     */
    @Test
    fun `android does not emit the dart-only codes`() {
        assertNull(SignerErrorCode.from("unsupported_platform"))
        assertNull(SignerErrorCode.from("unknown"))
    }

    @Test
    fun `every error code round-trips`() {
        for (code in SignerErrorCode.entries) {
            assertEquals(code, SignerErrorCode.from(code.wire))
        }
    }

    @Test
    fun `error code wire names are unique`() {
        assertEquals(SignerErrorCode.entries.size, SignerErrorCode.entries.map { it.wire }.toSet().size)
    }

    @Test
    fun `error codes are snake_case and methods are camelCase`() {
        // Not cosmetic: the two conventions differ on purpose, and mixing them
        // is a runtime-only failure on one platform.
        for (code in SignerErrorCode.entries) {
            assertFalse(code.wire.any { it.isUpperCase() }, "${code.wire} is not snake_case")
        }
        for (method in SignerMethod.entries) {
            assertFalse(method.wire.contains('_'), "${method.wire} is not camelCase")
        }
    }

    @Test
    fun `signer errors carry a typed code`() {
        val error = SecureKeyStore.SignerError(SignerErrorCode.KEY_NOT_FOUND, "nope")
        assertEquals(SignerErrorCode.KEY_NOT_FOUND, error.code)
        assertEquals("key_not_found", error.code.wire)
        assertNotNull(error.message)
    }
}

File: test/wire_contract_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_sign_keypair/flutter_sign_keypair.dart';

/// The wire contract, pinned.
///
/// Strings cross the method channel because that is all a channel can carry.
/// The mapping is duplicated once per language — Dart here, Kotlin in
/// `Wire.kt`, Swift in `Wire.swift` — and duplication is exactly what drifts.
/// These vectors are asserted verbatim in all three suites, so a rename on one
/// side fails a test on that side instead of failing a device at runtime.
///
/// If you change a string here you must change it in `Wire.kt` and `Wire.swift`
/// and in `WireContractTest.kt` and `WireContractTests.swift`. That is the
/// point: five deliberate edits, not one silent one.
void main() {
  group('KeyBacking wire mapping', () {
    // The canonical tags. Kotlin emits strongbox/tee/software; Swift emits
    // secure_enclave/keychain/software; Dart must understand all five.
    const Map<String, KeyBacking> vectors = <String, KeyBacking>{
      'strongbox': KeyBacking.strongBox,
      'secure_enclave': KeyBacking.secureEnclave,
      'tee': KeyBacking.trustedExecutionEnvironment,
      'keychain': KeyBacking.keychain,
      'software': KeyBacking.software,
    };

    vectors.forEach((String wire, KeyBacking expected) {
      test('"$wire" <-> ${expected.name}', () {
        expect(KeyBacking.fromWire(wire), expected);
        expect(expected.wireName, wire);
      });
    });

    test('every enum value round-trips — fails if a member is added blind', () {
      // Iterating `.values` is what makes this test survive a future edit: a new
      // member with a typo'd or duplicated wireName fails here rather than
      // silently mapping to `software` on a real device.
      for (final KeyBacking backing in KeyBacking.values) {
        expect(
          KeyBacking.fromWire(backing.wireName),
          backing,
          reason: '${backing.name} does not round-trip',
        );
      }
    });

    test('the vector table covers every enum value', () {
      expect(
        vectors.values.toSet(),
        KeyBacking.values.toSet(),
        reason:
            'a KeyBacking member was added without a wire vector — add it '
            'here AND in Wire.kt / Wire.swift and their tests',
      );
    });

    test('wire names are unique', () {
      final List<String> names = KeyBacking.values
          .map((KeyBacking b) => b.wireName)
          .toList();
      expect(names.toSet(), hasLength(names.length));
    });

    group('an unrecognised tag degrades to software, never upward', () {
      // This is the security-relevant direction. Under-reporting strength is
      // harmless; over-reporting is a false guarantee a caller would log.
      for (final String? tag in <String?>[
        null,
        '',
        'quantum',
        'STRONGBOX', // wrong case — must NOT match
        'strong_box', // plausible typo
        'secureEnclave', // camelCase instead of snake_case
        'tee ', // trailing space
        'Software',
      ]) {
        test('"$tag" -> software', () {
          final KeyBacking backing = KeyBacking.fromWire(tag);
          expect(backing, KeyBacking.software);
          expect(
            backing.isHardwareBacked,
            isFalse,
            reason: 'an unvalidated string must never yield a hardware claim',
          );
        });
      }
    });
  });

  group('KeyBacking.isHardwareBacked', () {
    test('only secure-element backings count', () {
      expect(KeyBacking.strongBox.isHardwareBacked, isTrue);
      expect(KeyBacking.secureEnclave.isHardwareBacked, isTrue);
      expect(KeyBacking.trustedExecutionEnvironment.isHardwareBacked, isTrue);
      // A keychain item is OS-protected but not held by a secure element.
      expect(KeyBacking.keychain.isHardwareBacked, isFalse);
      expect(KeyBacking.software.isHardwareBacked, isFalse);
    });

    test('every value has an explicit answer', () {
      // The getter is an exhaustive switch with no default, so this cannot
      // throw — but asserting it pins the intent if someone adds a default.
      for (final KeyBacking backing in KeyBacking.values) {
        expect(() => backing.isHardwareBacked, returnsNormally);
      }
    });
  });

  group('SignerErrorCode wire mapping', () {
    const Map<String, SignerErrorCode> vectors = <String, SignerErrorCode>{
      'key_not_found': SignerErrorCode.keyNotFound,
      'key_already_exists': SignerErrorCode.keyAlreadyExists,
      'hardware_unavailable': SignerErrorCode.hardwareUnavailable,
      'keystore_failure': SignerErrorCode.keystoreFailure,
      'user_authentication_required':
          SignerErrorCode.userAuthenticationRequired,
      'user_authentication_cancelled':
          SignerErrorCode.userAuthenticationCancelled,
      'key_invalidated': SignerErrorCode.keyInvalidated,
      'unsupported_platform': SignerErrorCode.unsupportedPlatform,
      'unknown': SignerErrorCode.unknown,
    };

    vectors.forEach((String wire, SignerErrorCode expected) {
      test('"$wire" <-> ${expected.name}', () {
        expect(SignerErrorCode.fromWire(wire), expected);
        expect(expected.wireName, wire);
      });
    });

    test('every enum value round-trips', () {
      for (final SignerErrorCode code in SignerErrorCode.values) {
        expect(SignerErrorCode.fromWire(code.wireName), code);
      }
    });

    test('the vector table covers every enum value', () {
      expect(vectors.values.toSet(), SignerErrorCode.values.toSet());
    });

    test('wire names are unique', () {
      final List<String> names = SignerErrorCode.values
          .map((SignerErrorCode c) => c.wireName)
          .toList();
      expect(names.toSet(), hasLength(names.length));
    });

    for (final String? tag in <String?>[null, '', 'nope', 'KEY_NOT_FOUND']) {
      test('an unrecognised code "$tag" becomes unknown', () {
        expect(SignerErrorCode.fromWire(tag), SignerErrorCode.unknown);
      });
    }

    test('the native sides only emit codes this side understands', () {
      // Kotlin's ErrorCode and Swift's ErrorCode both have exactly these
      // seven. `unsupported_platform` and `unknown` are Dart-side only:
      // the first comes from MissingPluginException, the second is the
      // catch-all. Pinned so adding a native code without a Dart member is
      // caught here.
      //
      // `key_invalidated` is emitted by Android only — iOS destroys a
      // `.biometryCurrentSet` key outright, so the same event surfaces there as
      // `key_not_found` on the user-present alias while the ambient one still
      // resolves. Listed anyway: a code either side can emit must have a Dart
      // member, and one-platform codes are the ones most likely to be missed.
      const Set<String> nativeCodes = <String>{
        'key_not_found',
        'key_already_exists',
        'hardware_unavailable',
        'keystore_failure',
        'user_authentication_required',
        'user_authentication_cancelled',
        'key_invalidated',
      };
      for (final String code in nativeCodes) {
        expect(
          SignerErrorCode.fromWire(code),
          isNot(SignerErrorCode.unknown),
          reason: 'native emits "$code" but Dart has no member for it',
        );
      }
      expect(
        SignerErrorCode.values
            .where((SignerErrorCode c) => c != SignerErrorCode.unknown)
            .where(
              (SignerErrorCode c) => c != SignerErrorCode.unsupportedPlatform,
            )
            .map((SignerErrorCode c) => c.wireName)
            .toSet(),
        nativeCodes,
      );
    });
  });

  group('SignerMethod wire mapping', () {
    const Map<String, SignerMethod> vectors = <String, SignerMethod>{
      'capabilities': SignerMethod.capabilities,
      'generateKey': SignerMethod.generateKey,
      'getKey': SignerMethod.getKey,
      'sign': SignerMethod.sign,
      'deleteKey': SignerMethod.deleteKey,
    };

    vectors.forEach((String wire, SignerMethod expected) {
      test('"$wire" <-> ${expected.name}', () {
        expect(expected.wireName, wire);
      });
    });

    test('the vector table covers every enum value', () {
      expect(vectors.values.toSet(), SignerMethod.values.toSet());
    });

    test('wire names are unique', () {
      final List<String> names = SignerMethod.values
          .map((SignerMethod m) => m.wireName)
          .toList();
      expect(names.toSet(), hasLength(names.length));
    });

    test('method names are camelCase, unlike the snake_case tags', () {
      // Not cosmetic: the Kotlin and Swift sides parse these verbatim, and
      // `generate_key` vs `generateKey` is a runtime-only failure on one
      // platform. Pinned so the convention cannot drift halfway.
      for (final SignerMethod method in SignerMethod.values) {
        expect(
          method.wireName,
          isNot(contains('_')),
          reason: '${method.name} broke the camelCase convention',
        );
      }
    });
  });

  group('SecureSignerException', () {
    test('keeps the raw code when it recognises it', () {
      final SecureSignerException e = SecureSignerException.fromWire(
        'key_not_found',
        'nope',
      );
      expect(e.code, SignerErrorCode.keyNotFound);
      expect(e.rawCode, 'key_not_found');
    });

    test('preserves an unrecognised raw code for diagnosis', () {
      final SecureSignerException e = SecureSignerException.fromWire(
        'some_future_native_code',
        'nope',
      );
      expect(e.code, SignerErrorCode.unknown);
      expect(
        e.rawCode,
        'some_future_native_code',
        reason: 'flattening the raw code away makes the failure undiagnosable',
      );
      expect(e.toString(), contains('some_future_native_code'));
    });

    test('a null wire code becomes unknown without throwing', () {
      final SecureSignerException e = SecureSignerException.fromWire(null, 'x');
      expect(e.code, SignerErrorCode.unknown);
      expect(e.rawCode, 'unknown');
    });

    test('toString names the enum, not the raw string, for known codes', () {
      expect(
        SecureSignerException(
          SignerErrorCode.hardwareUnavailable,
          'no secure element',
        ).toString(),
        'SecureSignerException(hardwareUnavailable): no secure element',
      );
    });
  });
}
